File: 763.PartitionLabels.py
from typing import List


class Solution:
    def partitionLabels(self, S: str) -> List[int]:
        s = S
        data = {}
        for i, v in enumerate(s):
            data[v] = i
        start = 0
        end = 0
        result: List[int] = []
        for i, v in enumerate(s):
            end = max(end, data[v])
            print("start", start)
            if end == i:
                result.append(end - start + 1)
                start = end + 1
        return result


if __name__ == "__main__":
    # Input: S = "ababcbacadefegdehijhklij"
    # Output: [9,7,8]
    # Explanation:
    # The partition is "ababcbaca", "defegde", "hijhklij".
    t = "ababcbacadefegdehijhklij"
    s = Solution()
    print(s.partitionLabels(t))

File: strings/763.PartitionLabels.go
package main

import "fmt"

func partitionLabels(S string) []int {
	s := S
	result := []int{}
	data := make([]int, 128)
	for i, v := range s {
		data[int(v)] = i
	}
	start := 0
	end := 0
	for i, v := range s {
		end = max(end, data[v])
		if end == i {
			result = append(result, end-start+1)
			start = end + 1
			end = 0
		}
	}
	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	// Input: S = "ababcbacadefegdehijhklij"
	// Output: [9,7,8]
	// Explanation:
	// The partition is "ababcbaca", "defegde", "hijhklij".
	t := "ababcbacadefegdehijhklij"
	fmt.Println(partitionLabels(t))
}
